srml/metadata/src/lib.rs

```diff
@@ -314,22 +319,30 @@ pub struct OuterDispatchCall {
 	pub index: u16,
 }
 
-/// All metadata about an runtime module.
-#[derive(Clone, PartialEq, Eq, Encode)]
+/// The metadata of a runtime.
+/// It is prefixed by a version ID encoded/decoded through
+/// the enum nature of `RuntimeMetadata`.
+#[derive(Eq, Encode, PartialEq)]
 #[cfg_attr(feature = "std", derive(Decode, Debug, Serialize))]
-pub struct RuntimeModuleMetadata {
-	pub prefix: DecodeDifferentStr,
-	pub module: DecodeDifferent<FnEncode<ModuleMetadata>, ModuleMetadata>,
-	pub storage: Option<DecodeDifferent<FnEncode<StorageMetadata>, StorageMetadata>>,
-}
+pub struct RuntimeMetadata(pub u32, pub RuntimeMetadataV1);
 
-/// The metadata of a runtime.
+/// The metadata of a runtime version 1.
 #[derive(Eq, Encode, PartialEq)]
 #[cfg_attr(feature = "std", derive(Decode, Debug, Serialize))]
-pub struct RuntimeMetadata {
-	pub outer_event: OuterEventMetadata,
+pub struct RuntimeMetadataV1 {
 	pub modules: DecodeDifferentArray<RuntimeModuleMetadata>,
-	pub outer_dispatch: OuterDispatchMetadata,
+}
+
+/// All metadata about an runtime module.
+#[derive(Clone, PartialEq, Eq, Encode)]
+#[cfg_attr(feature = "std", derive(Decode, Debug, Serialize))]
+pub struct RuntimeModuleMetadata {
+	pub name: DecodeDifferentStr,
+	pub prefix: DecodeDifferentStr,
+	pub storage: Option<DFn<StorageMetadata>>,
+	pub call: DFn<CallMetadata>,
+	pub outer_dispatch: DecodeDifferent<FnEncode<Option<OuterDispatchCall>>, Option<OuterDispatchCall>>,
+	pub event: DecodeDifferent<FnEncode<FnEncode<&'static [EventMetadata]>>, Vec<EventMetadata>>,
```

Comment:
Why don't we use an `Option` here? If we have no event, we can just use `None` and don't need to call some default function.